### ISSUE TITLE
Entity Ruler Speed with very large Vocabularies.

### ISSUE THREAD

**Issue by christiannawroth** [explosion/spaCy#5999]
Dear Team,

EntityRuler with very large vocabularies:

I understood the speed issues you mention in your documentation when adding large numbers of phrases.

However my main speed issue is when processing the documents after the phrases have been added.

I process relatively small documents (250 words) with a quite a large pattern vocabulary of 700.000 patterns. Compared to only using a statistical model (SciSpaCy) the entity ruler is really slow (~50 - 100 times, manually measured).

Could you give some advice (here or in the documentation) regarding EntityRulers speed in relation to the vocabulary size?

Thank you and all the best

Chris

## Info about spaCy

* **spaCy version:** 2.3.2
* **Platform:** Windows-10-10.0.18362-SP0
* **Python version:** 3.8.1

**Comment by adrianeboyd** [comment]
In the worst case, the `EntityRuler` is basically matching each token in the text against 700K patterns, which is going to be a lot more individual calculations than the statistical model uses to predict the label for a single token.

The actual runtime of the `EntityRuler` depends on a number of factors like the kind of patterns (token patterns vs. phrase patterns), how much overlap there is across the patterns (how many patterns start with the same token(s)), whether many patterns start with frequent tokens like `the`, and how many full and partial matches there are in your texts.

If you're using phrase patterns, the runtime for the `PhraseMatcher` is discussed in more detail here: https://explosion.ai/blog/spacy-v2-2#faster-phrasematcher and https://github.com/explosion/spaCy/pull/4309. Our observation was that runtime of the `PhraseMatcher` roughly doubles for every order of magnitude increase in the number of patterns, but there are so many factors that you may not see this exact behavior in your task.

Basically, it's hard to get around the fact that 700K patterns is a lot of patterns. You might try training a statistical NER model with the data labeled by your `EntityRuler` to see how close it can get to the same performance. You might also be able to combine a statistical NER model that performs well for frequent patterns with an `EntityRuler` with a smaller number of infrequent or particularly crucial patterns for your task.
